tests/unit_tests.cpp

```diff
@@ -0,0 +1,245 @@
+  cl::sycl::queue queue;
+  cl::sycl::buffer<int, 1> buf{cl::sycl::range<1>(num_threads)};
+  queue.submit([&](cl::sycl::handler& cgh) {
+    auto acc = buf.get_access<cl::sycl::access::mode::discard_write>(cgh);
+    cgh.parallel_for<class basic_parallel_for>(cl::sycl::range<1>(num_threads),
+      [=](cl::sycl::item<1> tid) {
+        acc[tid] = tid[0];
+      });
+    });
+  auto acc = buf.get_access<cl::sycl::access::mode::read>();
+  for(int i = 0; i < num_threads; ++i) {
+    BOOST_REQUIRE(acc[i] == i);
+  }
+}
+
+BOOST_AUTO_TEST_CASE(basic_parallel_for_nd) {
+  constexpr size_t num_threads = 128;
+  constexpr size_t group_size = 16;
+  cl::sycl::queue queue;
+  cl::sycl::buffer<int, 1> buf{cl::sycl::range<1>(num_threads)};
+  queue.submit([&](cl::sycl::handler& cgh) {
+    auto acc = buf.get_access<cl::sycl::access::mode::discard_write>(cgh);
+    cl::sycl::nd_range<1> kernel_range{cl::sycl::range<1>(num_threads),
+    cl::sycl::range<1>(group_size)};
+    cgh.parallel_for<class basic_parallel_for_nd>(kernel_range,
+      [=](cl::sycl::nd_item<1> tid) {
+        acc[tid.get_global()[0]] = tid.get_group(0);
+      });
+  });
+  auto acc = buf.get_access<cl::sycl::access::mode::read>();
+  for(int i = 0; i < num_threads; ++i) {
+    BOOST_REQUIRE(acc[i] == i / group_size);
+  }
+}
+
+BOOST_AUTO_TEST_CASE(hierarchical_dispatch) {
+  constexpr size_t local_size = 256;
+  constexpr size_t global_size = 1024;
+
+  cl::sycl::queue queue;
+  std::vector<int> host_buf;
+  for(size_t i = 0; i < global_size; ++i) {
+    host_buf.push_back(static_cast<int>(i));
+  }
+
+  {
+    cl::sycl::buffer<int, 1> buf{host_buf.data(), host_buf.size()};
+    queue.submit([&](cl::sycl::handler& cgh) {
+      auto acc = buf.get_access<cl::sycl::access::mode::read_write>(cgh);
+      cgh.parallel_for_work_group<class hierarchical_dispatch_reduction>(
+        cl::sycl::range<1>{global_size / local_size},
+        cl::sycl::range<1>{local_size},
+        [=](cl::sycl::group<1> wg) {
+          int scratch[local_size];
+          wg.parallel_for_work_item([&](cl::sycl::h_item<1> item) {
+            scratch[item.get_local_id()[0]] = acc[item.get_global_id()];
+          });
+          for(size_t i = local_size/2; i > 0; i /= 2) {
+            wg.parallel_for_work_item([&](cl::sycl::h_item<1> item) {
+              const size_t lid = item.get_local_id()[0];
+              if(lid < i) scratch[lid] += scratch[lid + i];
+            });
+          }
+          wg.parallel_for_work_item([&](cl::sycl::h_item<1> item) {
+            const size_t lid = item.get_local_id()[0];
+            if(lid == 0) acc[item.get_global_id()] = scratch[0];
+          });
+        });
+    });
+  }
+
+  for(size_t i = 0; i < global_size / local_size; ++i) {
+    size_t expected = 0;
+    for(size_t j = 0; j < local_size; ++j) expected += i * local_size + j;
+    size_t computed = host_buf[i * local_size];
+    BOOST_TEST(computed == expected);
+  }
+}
+
+BOOST_AUTO_TEST_CASE(local_memory) {
```

Comment:
Perhaps we can rename this to `dynamic_local_memory`? The previous test `hierarchical_dispatch` uses (static) local memory as well. While `hierarchical_dispatch`  tests the source transformation to insert a `__shared__` attribute, this tests the local memory accessor mechanism.